text/chapter10.md

```diff
@@ -391,416 +437,935 @@ foreign import isUndefined :: forall a. Undefined a -> Boolean
+> import Effect.Class.Console
+> import Effect.Aff
+> :pa
+… launchAff_ do
+…   log "waiting"
+…   toAff $ sleep 300
+…   log "done waiting"
+
+waiting
+unit
+done waiting
 ```
 
-The foreign JavaScript module is straightforward, defining the `alert` function by assigning it to the `exports` variable:
+Note that asynchronous logging in the repl just waits to print until the entire block has finished executing. This code behaves more predictably when run with `spago test` where there is a slight delay *between* prints.
 
-```javascript
-"use strict";
+PureScript's `Promise` type also represents `async`/`await` functions, which are just syntactic sugar for promises. Here we use the `async` keyword for the last parameter:
 
-exports.alert = function(msg) {
-    return function() {
-        window.alert(msg);
-    };
+```js
+exports.diagonalAsync = delay => w => async h => {
+  await exports.sleep(delay);
+  return Math.sqrt(w * w + h * h);
 };
 ```
 
-The `alert` action is very similar to the `log` action from the `Effect.Console` module. The only difference is that the `alert` action uses the `window.alert` method, whereas the `log` action uses the `console.log` method. As such, `alert` can only be used in environments where `window.alert` is defined, such as a web browser.
+Since we're returning a `Number` we represent this type in the `Promise`:
 
-Note that, as in the case of `log`, the `alert` function uses a function of no arguments to represent the computation of type `Effect Unit`.
+```hs
+foreign import diagonalAsync :: Int -> Number -> Number -> Promise Number
+```
 
-The second effect defined in this chapter is the `Storage` effect, which is defined in the `Effect.Storage` module. It is used to indicate that a computation might read or write values using the Web Storage API.
+```text
+$ spago repl
 
-The `Effect.Storage` module defines two actions: `getItem`, which retrieves a value from local storage, and `setItem` which inserts or updates a value in local storage. The two functions have the following types:
+import Prelude
+import Test.Examples
+import Control.Promise
+import Effect.Class.Console
+import Effect.Aff
+> :pa
+… launchAff_ do
+…   res <- toAff $ diagonalAsync 300 3.0 4.0
+…   logShow res
+…
+unit
+5.0
+```
 
-```haskell
-foreign import getItem :: String -> Effect Foreign
+We can also wrap `Promise` with `Effect`:
 
-foreign import setItem :: String -> String -> Effect Unit
+```hs
+foreign import diagonalAsyncEffect :: Int -> Number -> Number -> Effect (Promise Number)
 ```
 
-Here is the corresponding JavaScript implementation of these functions in `Effect/Storage.js`:
+The corresponding JavaScript needs to return a function of zero arguments to represent the `Effect`:
 
 ```js
-exports.setItem = function(key) {
-    return function(value) {
-        return function() {
-            window.localStorage.setItem(key, value);
-        };
-    };
+exports.diagonalAsyncEffect = delay => w => h => async function() {
```

Comment:
Why doesn't arrow syntax work with async effect?

This produces a foreign syntax error:
```js
exports.diagonalAsyncEffect = delay => w => h => async () => {
```

Reply:
Probably because it’s a gap in language-javascript, the Haskell library we use to parse FFI files. I’d really prefer we stick to ES5 functions in the book though.

Reply:
I wouldn’t recommend using promises here, unless you want to introduce the aff-promise library, which is pretty much the only way of working with them safely. They don’t work for typed FP very well at all, for two reasons:
- flattening violates parametricity and the monad laws, if you were tempted to give them that instance,
- eagerness: once you have a Promise value, the effects are already executing, which violates purity.

Reply:
> unless you want to introduce the aff-promise library

Yep. That library is used in these examples. I'd like to cover FFI with promises for completeness. I'll include the two caveats you mentioned. **Could you provide an example for the issues with "parametricity and the monad laws"?** Would the eagerness issue be solved by using the [promises](https://pursuit.purescript.org/packages/purescript-promises/3.0.0) library? Maybe that library could be touched-on too.

Also, I'm planning on covering async effects in the previous chapter (#129), so readers should have the necessary background for `Aff`.

ES6 in the book seems like an even smaller issue than ES6 in the libraries (although I do understand the concerns about wanting to reach everyone without using babel). But for the book, are there any readers on the cutting-edge, installing the purs compiler on their system, but refusing to use a modern browser? To address the concern that readers will use the ES6 code in the examples in their products, while being unaware of compatibility issues, I included a section on ["Modern JavaScript Syntax"](https://github.com/purescript-contrib/purescript-book/blob/093716ab07327bead268edc929b99f5aaefe3518/text/chapter10.md#a-note-about-modern-javascript-syntax) with a warning on ES6 syntax.

Reply:
Oh ok, sorry. I should have read more carefully. These FFI declarations are unsafe as-is, though; whenever you import a `Promise` from the FFI, it needs to be wrapped in an `Effect`, as described at https://github.com/nwolverson/purescript-aff-promise#consuming-promises-promise-to-aff.

I wouldn't recommend the `promises` library, I think it has worse ergonomics than `aff` + `aff-promise` and I'm not convinced that the approach does fully address the eagerness problem.

I'm not worried about readers of the book not using modern browsers, I'm worried about readers of the book writing libraries which use modern syntax and then publishing those libraries, and people then depending on those libraries and, as a result, shipping modern syntax to browsers that don't support it.

Reply:
Parametricity says that if you have a type variable bound with a `forall`, like in `forall a. Array a -> Array a`, your function's implementation is not allowed to know anything about the type `a`. In particular, it's not allowed to behave differently depending on whatever type the caller chooses for `a`. That means we can deduce that for a function whose type is `forall a. Array a -> Array a`, every element of the result is an element of the input. Promises violate this by looking at the result of the function you pass to `then` and flattening it if it looks like a promise.

Reply:
> whenever you import a Promise from the FFI, it needs to be wrapped in an Effect

Great point. I'll fix these.

> I'm worried about readers of the book writing libraries which use modern syntax and then publishing those libraries

I included a statement warning against that:
> It is also recommended to avoid arrow functions in public libraries for the same reason.

This could easily be overlooked though. Perhaps I should add emphasis.

Here are [some ideas](https://github.com/purescript/registry/issues/24) on how we can be more serious about protecting against accidental shipment of ES6 code, rather than just omitting ES6 code from examples.

I was following until here:
> Promises violate this by looking at the result of the function you pass to then and flattening it if it looks like a promise.

I think another PS snippet would help clarify this this point.

Reply:
I mean, yeah, warning against it works, but it seems wrong to me to encourage people to write code in this style at all in an introductory resource, especially if it would be rejected by the registry. Isn't it easier to avoid the issue entirely by sticking to ES5?

As for the Promise issue, consider:

```purescript
module Main where

import Prelude
import Effect (Effect)
import Effect.Console (logShow)
import Effect.Promise (Promise)
import Effect.Promise as Promise
import Effect.Promise.Console as Promise.Console

main = Promise.runPromise logShow logShow do
  hello <- Promise.resolve (Promise.resolve "hello")
  hello' <- hello
  Promise.Console.log (hello' <> ", world")
```
In the line `hello <- Promise.resolve (...)` we create a `Promise (Promise String)`. The type system thinks `hello` is a `Promise String`, because `<-` only "unwraps" one layer at a time. However, at runtime, `hello` is a String, because of promise flattening. The next line tries to call `then` on it, which fails, because it's not a promise. Here's the console output:
```
TypeError: promise.then is not a function
    at Object.exports.thenImpl (/home/harry/code/ps-scratch/output/Effect.Promise/foreign.js:2:18)
    at /home/harry/code/ps-scratch/output/Effect.Promise/index.js:55:29
    at /home/harry/code/ps-scratch/output/Data.Function/index.js:16:24
    at /home/harry/code/ps-scratch/output/Main/index.js:11:70
```

Reply:
The link with the monad laws is that the left identity law says that
```
y <- pure x
f y
```
should be the same as `f x`. Since `Promise.resolve` is `pure` for promises, we see that the above code can be rewritten from
```
  hello <- Promise.resolve (Promise.resolve "hello")
  hello' <- hello
  Promise.Console.log (hello' <> ", world")
```
to
```
  hello' <- Promise.resolve "hello"
  Promise.Console.log (hello' <> ", world")
```
and by applying the same law again, to
```
  Promise.Console.log ("hello" <> ", world")
```
but these two things are not the same, as the original example throws the "then is not a function" error but this one doesn't.

Reply:
> I think another PS snippet would help clarify this this point.

Nate clarified this on slack by writing:
> It’s not possible to construct Promise (Promise a) because the runtime always flattens that to Promise a. That makes it not parametric.

The Nullable docs describe this in more detail: https://pursuit.purescript.org/packages/purescript-nullable/4.1.1/docs/Data.Nullable#t:Nullable

Reply:
Thanks for the explanation. Unfortunately, github didn't refresh this comment thread, otherwise I wouldn't have posted that last comment. I should have posted here immediately after getting the answer on Slack.

Reply:
No worries! :slightly_smiling_face:

Reply:
> it seems wrong to me to encourage people to write code in this style at all in an introductory resource, especially if it would be rejected by the registry. Isn't it easier to avoid the issue entirely by sticking to ES5?

I think introductory readers are much more likely to just be writing FFI code for their own projects, rather than publishing to the registry. When readers progress enough to become active contributes, the ES5-to-ES6 transition should be painless, especially with available tools.

The ES6 syntax is WAY nicer to use (at least in my opinion). I think one of the selling points of PS is painless FFI, and writing nested curried functions without arrow functions is more painful. It's another "Pyramid of Doom" as discussed in the "Callback Hell" section of Ch12. Btw, that chapter also needs to be updated to acknowledge that's a non-issue these days with modern JS.

Reply:
Ok, fair enough, I'm convinced. :+1:

Reply:
Is it possible to get into one of the bad situations you described above when using just the [`aff-promise`](https://pursuit.purescript.org/packages/purescript-aff-promise) package (what's featured in this chapter), instead of the [`promises`](https://pursuit.purescript.org/packages/purescript-promises) package (used for your examples)?

I'd like to include your cautionary examples in the `promises` readme, but it doesn't seem to fit very well in the book, and I haven't been able to develop a similar problematic example with `aff-promise`.

Reply:
> whenever you import a `Promise` from the FFI, it needs to be wrapped in an `Effect`

Fixed in https://github.com/purescript-contrib/purescript-book/pull/140/commits/03e2dc99db9e1e5e94f33b880a053a4bc3dbda67

I was also just about to open an issue to document flattening, but found you already filed https://github.com/Thimoteus/purescript-promises/issues/12

If there are no gotchas with `aff-promise` that need to be documented, shall we mark this conversation as resolved?

Reply:
Sounds good. I am not aware of any similar gotchas with `aff-promise`.